File: wallet-kit/src/main/java/bitcoin/walllet/kit/network/message/GetDataMessage.java
package bitcoin.walllet.kit.network.message;

import java.io.ByteArrayInputStream;
import java.io.IOException;

import bitcoin.walllet.kit.common.io.BitcoinInput;
import bitcoin.walllet.kit.common.io.BitcoinOutput;
import bitcoin.walllet.kit.struct.InvVect;

/**
 * Build P2P message:
 * https://en.bitcoin.it/wiki/Protocol_documentation#Message_structure
 *
 * @author liaoxuefeng
 */
public class GetDataMessage extends Message {

    InvVect[] inventory; // byte[36]

    public GetDataMessage(int type, byte[][] hashes) {
        super("getdata");
        this.inventory = new InvVect[hashes.length];
        for (int i = 0; i < this.inventory.length; i++) {
            InvVect iv = new InvVect();
            iv.type = type;
            iv.hash = hashes[i];
            this.inventory[i] = iv;
        }
    }

    public GetDataMessage(byte[] payload) throws IOException {
        super("getdata");
        try (BitcoinInput input = new BitcoinInput(new ByteArrayInputStream(payload))) {
            long count = input.readVarInt(); // do not store count
            this.inventory = new InvVect[(int) count];
            for (int i = 0; i < this.inventory.length; i++) {
                this.inventory[i] = new InvVect(input);
            }
        }
    }

    @Override
    protected byte[] getPayload() {
        BitcoinOutput output = new BitcoinOutput().writeVarInt(this.inventory.length);
        for (int i = 0; i < this.inventory.length; i++) {
            output.write(this.inventory[i].toByteArray());
        }
        return output.toByteArray();
    }

    @Override
    public String toString() {
        return "GetDataMessage(count=" + this.inventory.length + ")";
    }

}

File: wallet-kit/src/main/kotlin/bitcoin/wallet/kit/messages/GetDataMessage.kt
package bitcoin.wallet.kit.messages

import bitcoin.wallet.kit.models.InventoryItem
import bitcoin.walllet.kit.io.BitcoinInput
import bitcoin.walllet.kit.io.BitcoinOutput
import java.io.ByteArrayInputStream
import java.io.IOException

/**
 * GetData Message
 *
 *  Size        Field           Definition
 *  ====        =====           ==========
 *  VarInt      Count           Number of inventory items
 *  Variable    InventoryItem   One or more inventory items
 */
class GetDataMessage : Message {

    lateinit var inventory: Array<InventoryItem> // byte[36]

    constructor(inventory: Array<InventoryItem>) : super("getdata") {
        this.inventory = inventory
    }

    constructor(type: Int, hashes: Array<ByteArray>) : super("getdata") {
        inventory = Array(hashes.size) { i ->
            val iv = InventoryItem()
            iv.type = type
            iv.hash = hashes[i]
            iv
        }
    }

    @Throws(IOException::class)
    constructor(payload: ByteArray) : super("getdata") {
        BitcoinInput(ByteArrayInputStream(payload)).use { input ->
            val count = input.readVarInt() // do not store count
            inventory = Array(count.toInt()) {
                InventoryItem(input)
            }
        }
    }

    override fun getPayload(): ByteArray {
        val output = BitcoinOutput().writeVarInt(inventory.size.toLong())
        for (i in inventory.indices) {
            output.write(inventory[i].toByteArray())
        }
        return output.toByteArray()
    }

    override fun toString(): String {
        return "GetDataMessage(count=${inventory.size})"
    }
}
